homeassistant/components/zwave/migration.py

```diff
@@ -0,0 +1,166 @@
+    migration_handler.add_entity_value(entity_id, entity_values)
+
+
+async def async_get_migration_data(
+    hass: HomeAssistant, config_entry: ConfigEntry
+) -> dict[str, ZWaveMigrationData]:
+    """Return Z-Wave migration data."""
+    migration_handler = get_legacy_zwave_migration(hass)
+    return await migration_handler.get_data(config_entry)
+
+
+@singleton(LEGACY_ZWAVE_MIGRATION)
+@callback
+def get_legacy_zwave_migration(hass: HomeAssistant) -> LegacyZWaveMigration:
+    """Return legacy Z-Wave migration handler."""
+    return LegacyZWaveMigration(hass)
+
+
+class LegacyZWaveMigration:
+    """Handle the migration from zwave to ozw and zwave_js."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Set up migration instance."""
+        self._hass = hass
+        self._store = Store(hass, STORAGE_VERSION, STORAGE_KEY)
+        self._data: dict[str, dict[str, ZWaveMigrationData]] = defaultdict(dict)
+
+    async def load_data(self) -> None:
+        """Load Z-Wave migration data."""
+        stored = cast(dict, await self._store.async_load())
+        if stored:
+            self._data = stored
+
+    @callback
+    def save_data(
+        self, config_entry_id: str, entity_id: str, data: ZWaveMigrationData
+    ) -> None:
+        """Save Z-Wave migration data."""
+        self._data[config_entry_id][entity_id] = data
+        self._store.async_delay_save(self._data_to_save, STORAGE_WRITE_DELAY)
+
+    @callback
+    def _data_to_save(self) -> dict[str, dict[str, ZWaveMigrationData]]:
+        """Return data to save."""
+        return self._data
+
+    @callback
+    def add_entity_value(
+        self,
+        entity_id: str,
+        entity_values: ZWaveDeviceEntityValues,
+    ) -> None:
+        """Add info for one entity and Z-Wave value."""
+        ent_reg = async_get_entity_registry(self._hass)
+        dev_reg = async_get_device_registry(self._hass)
+
+        node = entity_values.primary.node
+        entity_entry = ent_reg.async_get(entity_id)
+        assert entity_entry
+        device_identifier, _ = node_device_id_and_name(
+            node, entity_values.primary.instance
+        )
+        device_entry = dev_reg.async_get_device({device_identifier}, set())
+        assert device_entry
+
+        # Normalize unit of measurement.
+        unit = entity_entry.unit_of_measurement
+        unit = unit and unit.lower()
+        if unit == "":
+            unit = None
+
+        data: ZWaveMigrationData = {
+            "node_id": node.node_id,
+            "node_instance": entity_values.primary.instance,
+            "command_class": entity_values.primary.command_class,
+            "command_class_label": entity_values.primary.label,
+            "value_index": entity_values.primary.index,
+            "device_id": device_entry.id,
+            "domain": entity_entry.domain,
+            "entity_id": entity_id,
+            "unique_id": entity_entry.unique_id,
+            "unit_of_measurement": unit,
```

Comment:
Is it necessary to store this one as we can also fetch it from the entity registry ?

Reply:
I guess it is easier to do it here, so keep it in.